packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesAppLoader.m

```diff
@@ -0,0 +1,241 @@
+  }
+  [_arrayLock unlock];
+}
+
+# pragma mark - internal
+
+- (BOOL)_shouldStartLoadingUpdate:(EXUpdatesUpdate *)updateManifest
+{
+  EXUpdatesAppController *controller = [EXUpdatesAppController sharedInstance];
+  return [controller.selectionPolicy shouldLoadNewUpdate:updateManifest withLaunchedUpdate:controller.launchedUpdate];
+}
+
+- (void)_finishWithError:(NSError *)error
+{
+  if (_errorBlock) {
+    _errorBlock(error);
+  }
+  [self _reset];
+}
+
+- (void)_finish
+{
+  EXUpdatesDatabase *database = [EXUpdatesAppController sharedInstance].database;
+  dispatch_async(database.databaseQueue, ^{
+    [self->_arrayLock lock];
+    for (EXUpdatesAsset *existingAsset in self->_existingAssets) {
+      NSError *error;
+      BOOL existingAssetFound = [database addExistingAsset:existingAsset toUpdateWithId:self->_updateManifest.updateId error:&error];
+      if (!existingAssetFound) {
+        // the database and filesystem have gotten out of sync
+        // do our best to create a new entry for this file even though it already existed on disk
+        NSData *contents = [NSData dataWithContentsOfURL:[[EXUpdatesAppController sharedInstance].updatesDirectory URLByAppendingPathComponent:existingAsset.filename]];
+        existingAsset.contentHash = [EXUpdatesUtils sha256WithData:contents];
+        existingAsset.downloadTime = [NSDate date];
+        [self->_finishedAssets addObject:existingAsset];
+      }
+      if (error) {
+        NSLog(@"Error searching for existing asset in DB: %@", error.localizedDescription);
+      }
+    }
+    NSError *assetError;
+    [database addNewAssets:self->_finishedAssets toUpdateWithId:self->_updateManifest.updateId error:&assetError];
+    if (assetError) {
+      [self->_arrayLock unlock];
+      dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        [self _finishWithError:assetError];
+      });
+      return;
+    }
+
+    if (![self->_erroredAssets count]) {
+      NSError *updateReadyError;
+      [database markUpdateReadyWithId:self->_updateManifest.updateId error:&updateReadyError];
+      if (updateReadyError) {
+        [self->_arrayLock unlock];
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+          [self _finishWithError:updateReadyError];
+        });
+        return;
+      }
+    }
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+      if (self->_erroredAssets.count) {
+        if (self->_errorBlock) {
+          self->_errorBlock([NSError errorWithDomain:kEXUpdatesAppLoaderErrorDomain
+                                                code:1012
+                                            userInfo:@{NSLocalizedDescriptionKey: @"Failed to load all assets"}]);
+        }
+      } else {
+        if (self->_successBlock) {
+          self->_successBlock(self->_updateManifest);
+        }
+      }
+
+      [self->_arrayLock unlock];
+      [self _reset];
+    });
+  });
```

Comment:
A lot of this code should be fine to run on the database queue. We might need to lock _erroredAssets and reading _errorBlock, but I think those are protected by _arrayLock. I am a little worried about holding _arrayLock when calling the blocks (doesn't seem like we need to lock anything so I might write this as:

```
if we want to run the error block
  errorBlock = self->_errorBlock
else if we want to run the success block
  successBlock = self->_successBlock
release lock

dispatch to queue that the caller wants to be called on (so we get off the DB queue too) {
  if (errorBlock) {
    errorBlock(...);
  } else if (successBlock) {
    successBlock(...);
  }

  reset
}